<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Javier Costas" />

<meta name="date" content="2018-06-13" />

<title>The competitivePRS user guide</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">The competitivePRS user guide</h1>
<h4 class="author"><em>Javier Costas</em></h4>
<h4 class="date"><em>2018-06-13</em></h4>



<div id="table-of-contents" class="section level2">
<h2>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#intro">Introduction</a></li>
<li><a href="#install">Installing the competitivePRS package</a></li>
<li><a href="#workflow">Workflow</a><br />
3.1 <a href="#map">From genes to SNPs</a><br />
3.2 <a href="#DR2">Calculation of PRS and statistical analysis by logistic regression</a><br />
3.3 <a href="#perm">Generation of random SNP sets matching the actual SNP set by some genomic properties</a><br />
3.4 <a href="#sign">Estimation of significance by permutation</a><br />
3.5 <a href="#plot">Visual inspection of results</a><br />
</li>
<li><a href="#ref">References</a></li>
<li><a href="#email">Contact</a></li>
</ol>
</div>
<div id="introduction" class="section level2">
<h2>1. Introduction <a name="intro"></a></h2>
<p>Genome-wide association studies (GWAs) revealed that the genetic basis of complex diseases is highly polygenic, i.e., many common single nucleotide variants (SNPs) along the genome contribute to the disease, each one conferring a very low risk. As a consequence, methods to estimate the collective effect of SNPs <em>in masse</em> on disease risk have been developed. One popular method estimates risk alleles and effects in GWAS discovery samples of large size and calculates a polygenic risk score (PRS) in target GWAS samples as the sum of risk alleles weighted by their effects (logarithm of OR), as determined in the GWAS discovery sample. See <a href="#ref">Wray et al. (2014)</a> for an explanation of the method.</p>
<p>The PRS may be based on all SNPs from a discovery GWAS data set, those under a <em>P</em> value threshold, or SNPs selected by biological criteria, such as those in genes belonging to a specific gene set (for instance, genes involved in stress-response) or genomic feature (for instance, SNPs in active enhancers in the relevant tissue for the disease). As a consequence of the high polygenicity of most complex diseases, it is expected that a large proportion of SNP sets with enough size will be associated with the disease.</p>
<p>The <code>competitivePRS</code> R package is a tool for the estimation of PRS of a SNP set, and calculation of its significance in a target case-control study based on comparison of the Nagelkerke’s pseudo-R<sup>2</sup> of the actual SNP set to random permutations of SNP sets matching the actual SNP set by a series of properties such as linkage disequilibrium (LD) pattern, minor allele frequency (MAF), distance to transcription start sites, and gene density. The scientific question to answer is not if the PRS based on the SNP set is associated with disease status but if it is more associated than a random SNP set with similar properties of LD, MAF and position in relation to genes. Because of that, it is considered a competitive test.</p>
</div>
<div id="installing-the-competitiveprs-package" class="section level2">
<h2>2. Installing the competitivePRS package <a name="install"></a></h2>
<p><code>competitivePRS</code> is distributed as a source code at GitHub. Therefore, the best way to install it is using the <code>devtools</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;costasjavier/competitivePRS&quot;</span>)</code></pre></div>
<p>Please, note that this process may take several minutes, as a large dataset has to be downloaded. The dependencies of <code>competitivePRS</code> are <code>GenomicRanges</code>, to map SNPs to genes; and <code>rms</code> to perform logistic regression analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)
<span class="kw">library</span>(rms)
<span class="kw">library</span>(competitivePRS)</code></pre></div>
<p>In addition, <strong>PLINK 1.9</strong> (<a href="#ref">Chang et al., 2015</a>) has to be installed in the PATH. The <strong>PLINK</strong> executables must be downloaded from the <strong><a href="https://www.cog-genomics.org/plink2/">PLINK</a></strong> site.</p>
</div>
<div id="workflow" class="section level2">
<h2>3. Workflow <a name="workflow"></a></h2>
<p>For use of <code>competitivePRS</code>, the user needs a summary statistic for a GWAS discovery sample, one or several SNP (or gene) sets to test, and genotyping data along the genome for a target case-control sample. The genotyping data have to be in the <strong>PLINK</strong> binary format. This common format stores the data in three files, with the extensions <em>“bim”</em>, <em>“fam”</em>, and <em>“bed”</em>, representing SNPs information, subjects information, and genotyping results, respectively.</p>
<p>This tutorial explains the different steps implemented in <code>competitivePRS</code> to obtain a permuted <em>P</em> value for the significance of the PRS based on a SNP set in comparison with genomic background, i.e. random SNP sets with similar properties of LD, MAF and location in reference to genes.</p>
<div id="from-genes-to-snps" class="section level3">
<h3>3.1. From genes to SNPs <a name="map"></a></h3>
<p>The <code>Map_genestoSNPs</code> function takes five arguments. The first argument, <code>geneset</code> is a list of genes using the HGNC symbol (such as SLC39A8) or NCBI gene ID (such as 20862) for identification, but never a combination of both. This list may be an array or a dataframe with just one column. For instance,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.gene.list &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;COMT&quot;</span>, <span class="st">&quot;PRODH&quot;</span>, <span class="st">&quot;TTC28&quot;</span>, <span class="st">&quot;MTMR3&quot;</span>, <span class="st">&quot;FFFF&quot;</span>)
my.gene.list</code></pre></div>
<pre><code>## [1] &quot;COMT&quot;  &quot;PRODH&quot; &quot;TTC28&quot; &quot;MTMR3&quot; &quot;FFFF&quot;</code></pre>
<p>Alternatively, the list may be read from a file. In this tutorial, we are going to use the example files included within the package. To retrieve the full path to the example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="st">&quot;exGenelist.txt&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;competitivePRS&quot;</span>)</code></pre></div>
<p>Now, read the example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.gene.list &lt;-<span class="st"> </span><span class="kw">read.table</span>(path, <span class="dt">header =</span> F)
<span class="kw">head</span>(my.gene.list)</code></pre></div>
<pre><code>##        V1
## 1    SYT1
## 2    GRM1
## 3    GRM5
## 4    DLG2
## 5 EPB41L1
## 6  GRIN2A</code></pre>
<p>The second argument is a string with the full <code>path.to.plink.files</code>, where the three files of the <strong>PLINK</strong> binary format are present. These files store the genotype data of the target sample file, the user’s data. To retrieve the path of the example data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;competitivePRS&quot;</span>)
path&lt;-<span class="kw">paste0</span>(path,<span class="st">&quot;/&quot;</span>)</code></pre></div>
<p>The third argument is a string indicating the name of the <strong>PLINK</strong> binary file, without extension. The four argument is an array of two numbers, indicating the length in kb of extra bases at the 5’ and 3’ edges of the exonic region of a gene to be considered as part of a gene, respectively. The final argument is a string indicating the name of the output file to write the list of genes not identified. This file will be written in the <code>path.to.plink.files</code> directory. A typical use may be:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_SNP_list &lt;-<span class="st"> </span><span class="kw">Map_genestoSNPs</span>(<span class="dt">geneset =</span> my.gene.list, <span class="dt">path.to.plink.files =</span> path, <span class="dt">bfile =</span> <span class="st">&quot;exPLINK&quot;</span>, <span class="dt">extra.kb =</span> <span class="kw">c</span>(<span class="dv">35</span>,<span class="dv">10</span>), <span class="dt">output =</span> <span class="st">&quot;my_output_1&quot;</span>)</code></pre></div>
<pre><code>## A total of 2 genes from gene set were not found (written to output file)</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(my_SNP_list)</code></pre></div>
<pre><code>##            V1
## 1 12:79252123
## 2 12:79253404
## 3 12:79271359
## 4 12:79275054
## 5 12:79291372
## 6 12:79294033</code></pre>
<p>In this case, the SNPs located between 35 kb upstream and 10 kb downstream of each gene from the supplied list are assigned to it. The default value is <code>extra.kb=c(0,0)</code>, to limit to SNPs to those between the first and last exonic bases of a gene. This function generates a dataframe of one column with the SNPs in the genotype data mapping on the genes in the supplied list , identified by chromosome and position (ex. 4:103188709). A message in the console indicates the number of genes not found. This problem usually originates by the use of different versions of the genome. <code>competitivePRS</code> uses the NCBI37.3 (hg19) gene location.</p>
</div>
<div id="calculation-of-prs-and-statistical-analysis-by-logistic-regression" class="section level3">
<h3>3.2. Calculation of PRS and statistical analysis by logistic regression <a name="DR2"></a></h3>
<p>This may be the next step after generation of a SNP set by the previous step or, alternatively, be the first step if the researcher begins the analysis with a SNP set selected by other way. In this case, the SNPs have to be identified by chr:position (ex. 4:103188709), and have to be a dataframe object of just one column. If there are additional columns, only the first column is used.</p>
<p>The required function, <code>Calculate_DR2Nagelkerke</code>, begins with a generation of new <strong>PLINK</strong> binary bfiles with the column <em>snpid</em> corresponding to <em>chr:position</em>, and removing all SNPs with the same <em>chr:position</em>, usually, duplicate SNPs with different IDs or triallelic SNPs. If the logical argument <code>amb.remove</code> is set to <code>TRUE</code>, SNPs with strand anbiguity (A&lt;-&gt;T and C&lt;-&gt;G) are removed. The new <strong>PLINK</strong> binary files are written to the <code>path.to.plink.files</code>, using the generic name <em>“tmp”</em>.</p>
<p>Then, the function calls <strong>PLINK</strong> to perform LD-clumping and scoring. The objective of LD-clumping (also known as informed pruning) is to generate a SNP subset with SNPs in approximate linkage equilibrium, choosing the most significant SNP in the GWAS discovery sample among those SNPs in LD. The clumping is done with the following arguments: <em>“–clump-kb 500 –clump-p1 1 –clump-p2 1 –clump-r2 0.1”</em> (see the <strong>PLINK</strong> manual for details). A <em>P<sub>threshold</sub></em> at GWAS discovery sample for inclusion of SNPs in the PRS calculation may be specified by the user. Taking into account the rationale under the permutation test of <code>competitivePRS</code>, a liberal <em>P<sub>threshold</sub></em> is more appropriate. If the <em>P<sub>threshold</sub></em> is very restrictive, the PRS from a gene set may be represented by very few SNPs and genes. In addition, if the GWAS discovery sample has a reasonable power, rejection of the null hypothesis under a very restrictive <em>P<sub>threshold</sub></em> is obvious and lack interest. Imaging a <em>P<sub>threshold</sub></em> = 1e<sup>-5</sup>, if any SNP from the gene set reach this <em>P</em> value, a PRS based on these SNPs is expected to explain more variance on the trait than any set of random SNPs. Testing this hypothesis is not particularly interesting. The default value of the <code>P.thershold</code> argument is 1, that is, inclusion of all the SNPs, irrespective of the <em>P</em> value.</p>
<p>Then, a logistic regression analysis, using the <code>rms</code> package, is carried out, allowing for the use of covariates. The percentage of SNPs from the PRS model that failed in a particular sample (i.e. missing genotypes) is included as a covariate by default. In addition, if the user chooses to calculate a better coefficient of determination than Nagelkerke’s pseudo-R<sup>2</sup> (as described in <a href="#ref">Lee et al., 2012</a>), its value is also included. In this case, an array indicating disease prevalence and proportion of cases in the target sample have to be passed to the <code>better.coef</code> argument.</p>
<p>Two data sets are read to a dataframe from file. The first one corresponds to data from summary statistics of the GWAS discovery sample, including a minimum of these columns <em>“snpid”</em>, <em>“a1”</em>, <em>“p”</em>,<em>“effect”</em>. If the logical argument <code>amb.remove</code> is set to <code>TRUE</code>, and additional column, <em>a2</em>, is also required.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path1 &lt;-<span class="st"> </span><span class="kw">paste</span>(path,<span class="st">&quot;exSummaryGWAS&quot;</span>,<span class="dt">sep =</span> <span class="st">&quot;/&quot;</span>)
ex_summGWAS &lt;-<span class="st"> </span><span class="kw">read.table</span>(path1, <span class="dt">header =</span> T)
<span class="kw">head</span>(ex_summGWAS)</code></pre></div>
<pre><code>##         snpid a1 a2      p       effect
## 1 1:100005477  A  G 0.9260 -0.001400981
## 2 1:100059035  T  C 0.3831 -0.009404080
## 3 1:100079760  T  C 0.1886  0.014504302
## 4 1:100081117  T  G 0.4249 -0.013004189
## 5 1:100086858  A  G 0.4394 -0.009202211
## 6 1:100107260  T  C 0.1019 -0.017400515</code></pre>
<p>The second one is a data set with information of covariates. This is an optional file, in case covariates are included in the logistic regression model. Typical covariates in this type of studies include sex, age, and main PCA (or MDS) dimensions to control for population substructure. The file has to include as many covariates as needed, with a header with covariate names. The covariates to include in the model have to be specified, therefore, this file may contain more covariates than the useful ones. The Dependent Variable, “PHENO”, is taken from the <strong>PLINK</strong> <em>“fam”</em> file. A “IID” column is required for sample identification.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path2 &lt;-<span class="st"> </span><span class="kw">paste</span>(path,<span class="st">&quot;exCovariates&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>)
Covariates &lt;-<span class="kw">read.table</span>(path2,<span class="dt">header =</span> T)
<span class="kw">head</span>(Covariates)</code></pre></div>
<pre><code>##      IID Sex        PCA1        PCA2
## 1 11_002   2 -0.20555026  0.15640545
## 2 11_029   2  0.28650108 -0.10795737
## 3 11_003   2 -0.17907436  0.22686469
## 4 11_022   1 -0.01054431 -0.13947130
## 5 11_007   2  0.03156429 -0.09405964
## 6 11_035   2  0.29811221  0.09599895</code></pre>
<p>For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PRS_analysis &lt;-<span class="st"> </span><span class="kw">Calculate_DR2Nagelkerke</span>(<span class="dt">SNPs.list =</span> my_SNP_list, <span class="dt">sumGWAS =</span> ex_summGWAS, <span class="dt">Cov =</span> Covariates,      <span class="dt">path.to.plink.files =</span> path, <span class="dt">bfile =</span> <span class="st">&quot;exPLINK&quot;</span>, <span class="dt">P.threshold =</span> <span class="fl">0.5</span>, <span class="dt">amb.remove =</span> T,
<span class="dt">cov.names =</span> <span class="kw">c</span>(<span class="st">&quot;Sex&quot;</span>,<span class="st">&quot;PCA2&quot;</span>), <span class="dt">better.coef =</span> <span class="kw">c</span>(<span class="fl">0.05</span>,<span class="fl">0.5</span>), <span class="dt">output.file =</span> <span class="st">&quot;my_output_2&quot;</span>)</code></pre></div>
<pre><code>## 
## Results written to file C:/Users/XCostas/Documents/R/win-library/3.2/competitivePRS/extdata/my_output_2</code></pre>
<p>The function returns a list of four elements, the increase in Nagelkerke’s pseudo-R<sup>2</sup> after adding the standardized PRS to a model including only covariates, an array with the SNPs present in the PRS model (needed as input of the <code>Generate_permutationSets</code> function), the summary of the fitted logistic regression model, for inspection by the user, and a better coefficient of determination (or “n.a.” if no requested).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PRS_analysis<span class="op">$</span><span class="st">`</span><span class="dt">Pseudo-R2</span><span class="st">`</span></code></pre></div>
<pre><code>##         R2 
## 0.01735782</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PRS_analysis<span class="op">$</span>Input_perm</code></pre></div>
<pre><code>##  [1] &quot;11:83204132&quot; &quot;11:83373309&quot; &quot;11:83479808&quot; &quot;11:83562296&quot; &quot;11:83581941&quot;
##  [6] &quot;11:83620215&quot; &quot;11:83634983&quot; &quot;11:83655564&quot; &quot;11:83728714&quot; &quot;11:83918339&quot;
## [11] &quot;11:84022936&quot; &quot;11:84192441&quot; &quot;11:84321748&quot; &quot;11:84379740&quot; &quot;11:84383445&quot;
## [16] &quot;11:84449873&quot; &quot;11:84572377&quot; &quot;11:84574040&quot; &quot;11:84642163&quot; &quot;11:85280558&quot;
## [21] &quot;11:88252273&quot; &quot;12:13723127&quot; &quot;12:13852050&quot; &quot;12:13870080&quot; &quot;12:13896895&quot;
## [26] &quot;12:13950577&quot; &quot;12:13968186&quot; &quot;12:14015706&quot; &quot;12:14065831&quot; &quot;12:79252123&quot;
## [31] &quot;12:79484134&quot; &quot;12:79592680&quot; &quot;12:79738884&quot; &quot;12:79785690&quot; &quot;16:10208218&quot;
## [36] &quot;16:10216343&quot; &quot;16:10225050&quot; &quot;16:10304290&quot; &quot;16:9850397&quot;  &quot;16:9873465&quot; 
## [41] &quot;16:9940293&quot;  &quot;6:146677281&quot; &quot;6:146680907&quot; &quot;6:146725969&quot; &quot;6:146726569&quot;
## [46] &quot;6:146766138&quot; &quot;11:84786132&quot; &quot;11:88669352&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PRS_analysis<span class="op">$</span>Logistic_model</code></pre></div>
<pre><code>## Logistic Regression Model
##  
##  rms::lrm(formula = formula.m1)
##  
##                       Model Likelihood     Discrimination    Rank Discrim.    
##                          Ratio Test           Indexes           Indexes       
##  Obs           100    LR chi2      4.22    R2       0.055    C       0.594    
##   1             50    d.f.            4    g        0.511    Dxy     0.188    
##   2             50    Pr(&gt; chi2) 0.3769    gr       1.666    gamma   0.189    
##  max |deriv| 0.004                         gp       0.104    tau-a   0.095    
##                                            Brier    0.241                     
##  
##            Coef     S.E.    Wald Z Pr(&gt;|Z|)
##  Intercept   0.1373  0.6566  0.21  0.8344  
##  st.score    0.2381  0.2072  1.15  0.2505  
##  miss      -18.4065 62.8014 -0.29  0.7695  
##  Sex        -0.0356  0.4083 -0.09  0.9306  
##  PCA2       -1.6669  1.3202 -1.26  0.2067  
## </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PRS_analysis<span class="op">$</span>Better_coef</code></pre></div>
<pre><code>##        R2 
## 0.0111257</code></pre>
<p>The Independent Variables in the model are <em>“st.score”</em>, corresponding to the standardized PRS; <em>“miss”</em>, corresponding to the <em>“missing genotypes”</em> covariate; and the remaining optional covariates.</p>
<p>In addition, these and additional data are stored in the output file to facilitate latter inspection of results.</p>
<p>It is important to note that the GWAS discovery sample and the GWAS target sample have to be concordant in terms of DNA strand to call alleles. Lack of concordance is easy to identify in some cases, as the existence of more than two alleles. This is shown in the <strong>PLINK</strong> output as a warning message with the number of skipped SNPs as a consequence of allele mismatches. But in some instances (A/T or C/G SNPs), the mismatch is undetectable. <code>competitivePRS</code> does not deal with this problem. It is responsibility of the user to check the two GWAS samples for strand concordance, flipping strand or removing ambiguous SNPs before the analysis.</p>
</div>
<div id="generation-of-random-snp-sets-matching-the-actual-snp-set-by-some-genomic-properties" class="section level3">
<h3>3.3. Generation of random SNP sets matching the actual SNP set by some genomic properties <a name="perm"></a></h3>
<p>In the example above, the PRS was not significant. But assuming it was, an additional question has to be solved. Is the PRS for this SNP set more explanative than random SNP sets along the genome? The <code>competitivePRS</code> package addresses this question by permutations. The first step is to create random SNP sets with the same size of the original one and matching the actual SNPs by several features such as LD pattern, MAF, gene density, or distance to the nearest transcription start site. This is achieved by the <code>Generate_permutationTest</code> function. This function generates random sets of SNPs approximately matching the actual SNP set in the above-mentioned properties, based on the 1000 genomes European population, taking data from <strong>SNPsnap</strong> (<a href="#ref">Pers et al., 2015</a>). The extended Major Histocompatibility Complex (chr6, from 25912984 to 33290793 in hg19) was excluded due to its complex linkage disequilibrium structure. For faster computation, the SNPs from <strong>SNPsnap</strong> were previously classified in sets of similar characteristics. Specifically, the SNPs were classified by deciles except for gene counts, due to the excess of SNPs with low gene counts. Thus, 80.3% of SNPs present 0-4 gene counts. The classes to match were 0, 1, 2, 3, 4, and two additional classes, evenly distributed. Those SNPs not present in the <strong>SNPsnap</strong> data set or without any matched SNP in the <strong>SNPsnap</strong> data set are included in each permutation, to get the same number of SNPs than in the original SNP set.</p>
<p>To run the function, five arguments have to be passed. <code>input.perm</code> is an array with the actual SNPs to match. This is the second element of the list generated by <code>Calculate_DR2Nagelkerke</code>, in our example, <code>PRS_analysis[[2]]</code>. The <code>n.perm</code> argument indicates the required number of permutations. <code>bfile</code> is an string with the name of the <strong>PLINK</strong> binary file (without extension) with actual data. The “tmp” name, corresponding to the bfile generated by the <code>Calculate_DR2Nagelkerke</code> function, is used by default. The other two arguments are in common to the <code>Calculate_DR2Nagelkerke</code> function, the summary statistics of the GWAS discovery sample, and the path to to the three <strong>PLINK</strong> binary files (<em>“.bed”</em>, <em>“.bim”</em>, <em>“.fam”</em>).</p>
<p>The function first creates a subset of the SNPsnap data set with SNPs shared with the summary GWAS and the target GWAS. Then, it presents a message with information on SNPs from the actual set to generate the PRS model not in the <strong>SNPsnap</strong> data set. If the proportion of SNPs absent from the <strong>SNPsnap</strong> data set is high, interpretation of the test may be flawed. Finally, it generates a dataframe with the random SNP sets matching the actual SNP set in the above-mentioned features. Each SNP set corresponds to a column of the dataframe.</p>
<p>In this example, 50 permutations are generated. This is a very low number, just to test the package. <code>n.perm</code> values of 1000-10000 seem reasonable to take conclusions in most cases.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">random_sets &lt;-<span class="st"> </span><span class="kw">Generate_permutationSets</span>(<span class="dt">input.perm =</span> PRS_analysis[[<span class="dv">2</span>]], <span class="dt">sumGWAS =</span> ex_summGWAS, <span class="dt">bfile =</span> <span class="st">&quot;tmp&quot;</span>, <span class="dt">path.to.plink.files =</span> path, <span class="dt">n.perm =</span> <span class="dv">50</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(random_sets)</code></pre></div>
<pre><code>## [1] 48 50</code></pre>
<p>The first lines and columns of the output looks like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">random_sets[<span class="dv">1</span><span class="op">:</span><span class="dv">9</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</code></pre></div>
<pre><code>##            X1          X2           X3          X4          X5          X6
## 1 3:139648874  1:96393689  22:32864879 13:46426764 4:183849072 12:14128322
## 2 2:165183257 18:59291851 12:130906479  4:78336408  8:74335491  8:74335491
## 3  3:94943493 11:23804832  11:23804832 4:161441599 11:83918339 11:28826913
## 4  4:21116890  16:7493024   3:60498172   9:9539114  3:54867313 6:124890305
## 5 9:109813447 13:27488025  10:64278634 7:132192949 13:67933131 22:44545365
## 6 14:47386441  2:57218736   2:57218736 14:47386441  2:57218736 4:130724902
## 7 11:84642163 11:84642163  11:84642163 11:84642163 11:84642163 11:84642163
## 8   9:9216473  5:58646112 11:132543770  2:80169400  4:62859511  10:9017944
## 9 11:83204132  5:84562600   3:17357820 3:162925738  1:49454215 11:83283859</code></pre>
</div>
<div id="estimation-of-significance-by-permutation" class="section level3">
<h3>3.4. Estimation of significance by permutation <a name="sign"></a></h3>
<p>Once the random sets of matched SNPs were generated, next step if estimation of significance by the <code>Calculate_permutationSignificance</code> function. Basically, the function first uses the <em>–score</em> option of <strong>PLINK</strong> to estimate PRS based on each random subset and, then, calculates the Nagelkerke’s pseudo-R<sup>2</sup> for each PRS to create a null distribution, including the same covariates in the logistic regression model as in the original analysis. The Nagelkerke’s pseudo-R<sup>2</sup> of the actual SNP set is compared against this distribution to obtain the permuted <em>P</em> value, following <a href="#ref">Phipson and Smyth (2010)</a>. The result is showed in the console and saved to a file for inspection. The values of Nagelkerke’s pseudo-R<sup>2</sup> for each permutation are also saved to generate a plot with the distribution (see below).</p>
<p>This function accepts eight arguments, most of them common to other functions of the <code>competitivePRS</code> package, such as <code>path.to.plink.files</code>, <code>sumGWAS</code>, <code>Cov</code>, <code>cov.names</code>, <code>bfile</code> or <code>outfile</code>, previously described. Two additional arguments are required. <code>DR2</code> is a number with the value of the Naglekerke’s pseudo-R<sup>2</sup> of PRS in the actual data set. This is the first element of the list generated by the <code>Calculate_DR2Nagelkerke</code> function of this package. The user can also pass this value by writing it. <code>Perm</code> is a dataframe with the permuted SNP sets, the object created by the <code>Generate_permutationTest</code> function.</p>
<p>Following the example,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">result_test &lt;-<span class="st"> </span><span class="kw">Calculate_permutationSignificance</span>(<span class="dt">path.to.plink.files =</span> path, <span class="dt">sumGWAS =</span> ex_summGWAS, 
<span class="dt">Perm =</span> random_sets, <span class="dt">Cov =</span> Covariates, <span class="dt">cov.names =</span> <span class="kw">c</span>(<span class="st">&quot;Sex&quot;</span>,<span class="st">&quot;PCA2&quot;</span>), <span class="dt">DR2 =</span> PRS_analysis[[<span class="dv">1</span>]], <span class="dt">outfile =</span> <span class="st">&quot;my_output_3&quot;</span>)</code></pre></div>
<pre><code>## The permuted P value is: 0.215686274509804</code></pre>
<p>The <strong>PLINK</strong> output is shown in the console. The output of the <code>Calculate_permutationSignificance</code> function is a list of two elements, the <em>P</em> value, and an array with the values of pseudo-R<sup>2</sup> of Nalgelkerke for each permutation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(result_test)</code></pre></div>
<pre><code>## List of 2
##  $ Permuted.P      : num 0.216
##  $ DR2.permutations: num [1:50] 0.002308 0.006951 0.000202 0.000549 0.040321 ...</code></pre>
</div>
<div id="visual-inspection-of-results" class="section level3">
<h3>3.5. Visual inspection of results <a name="plot"></a></h3>
<p>Finally, <code>competitivePRS</code> includes a simple function to graphically represent the distribution of the increase in Nagelkerke’s pseudo-R<sup>2</sup> for the random SNP sets. A vertical red line with the value for the actual SNP set is also shown. The function, <code>Generate_Plot</code> needs two arguments, the values of pseudo-R<sup>2</sup> of Nalgelkerke for each permutation (that is, the second element of the list generated by the <code>Calculate_permutationSignificance</code> function), and the value of the Nagelkerke’s pseudo-R<sup>2</sup> of PRS in the actual data set (the first element of the list generated by the <code>Calculate_DR2Nagelkerke</code> function).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Generate_Plot</span>(<span class="dt">permDR2 =</span> result_test[[<span class="dv">2</span>]], <span class="dt">DR2 =</span> PRS_analysis[[<span class="dv">1</span>]])</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAWlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6kNtmAABmADpmkJBmtrZmtv+QOgCQZgCQkGaQtpCQ2/+2ZgC2/7a2///bkDrb/7bb////AAD/tmb/25D//7b//9v////xXCPTAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAIKklEQVR4nO2daYOaSBRF7RjTJtHJtDNMN2r9/785tbIJ3iqqcHne88GF5QmnawNew0qRq6zuvQGPDgUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIMAiglaBWWt/Fd6aPJYRlBWcggAUBKAgAAU9FRQEoCDAAwpiGwSgIAAFASQKWvWPvigIhGMvBsKJFHTarr6XCidSkFKVbkLei4ST2wZlO5IuyCn69m9mOLGCam1np8779WdeOJmCdCu9cmbq+UVIsKDT9u2jVDihvVi5cEIFHXQFqzM7esmCDrYBOm2zhoui26Cdfc9ooZVoQee9q1wVBfVo96FamSJ03HAk3aOzD8eNHidm9vWiBZULJ7MXKxhOpqDz3p00vXsj/Vh0xkFZZnrhRLZBp23uybI2nFBBu2LhRAo67yOOMWA7JViQPk6FRagKh7KTx7SCBZnzZaAXC0cjmmripKPkXgzTaaamDmlfW9CrlyBdydafh2udfRWaqVdsg1T99qHLxfXhUGioJi97CBZk6o+pODwf1Kc3UDSCZp1RfIX0l1CCDvOvGirRJci3QRUeLuJwcnux62cUm7Hk9HBStCDMeY8aqBcXhI9oBbdBEcdiCh/RChbkybtu+AKC1IGXnntc7AMvPfe52If7H2o8FsN9eIDsjsfiohfLOtIYCsr57/AH4RZXVhN/48HboBLhRAqKONCKDidSkB8kL3E0L0PQee/MTJ2OTwknUtDppzvRscBAUUYvFkrQAmcUZQgyZxT1a5WXgydZkOvH8saJktugkuEoCIQTKghfeo4NJ1NQzKXnyHAiBS146VlGLxZx6dmVrjo5BU+GoIhLz1aQPRSZzPgULCji0rMR5NUkJVDJaIOiLj2/q+PGChpUxOvZHVIEQV67BHUSEKdw59S+qytp+YIFxWXaa0e6Ek7/669gQbkXnbvhRPZicckLceFECioZjoJAOHltUJn/hVLCBUV09JHhKAiEoyAQjoJAuKGgUfrr3SAPZN5v3LAEjb6NLrkM836DgqLWKpPaocoIWqoNyhFUeCMoCISjIBBuviDDV3ZnNugpB98SY2Vuy2i43BKULSji91NjFaJML0ZBFz9MQVNv15bM3AqxgtgGubexriUISj5eG++onlvQ5NvXcP3RaGO/O1yEgibm3U/QvBssvY6gmTdYmvv2dILm3h7nZQRN32BpmN3x0CwnKKIEleARz0lHgm+wVIInFlTo/zYBzyzoJryAoLxG9Cu/HR6QtS+lpKQGLbRMqZ9aZuWcoBR0m2Uo6EY/tczKOUEp6DbLUNCNfmqZlXOCUtBtlnlOQZKgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBgEKC6u49G5ovdfJjSUfj2KSbpEu545uj5tzCrowgc4O4OmxF86U3dX6c816/VKuEuxSPb475nH5NuIggl9Xg72XefOlNzYjjbheScF+j8TDK32MrfnMsRQT1dqH5krxjE3HcvISiOBmmWv91J0E/zMb7nKHmS29qRhw3L+Epp1Nh9Mc7tUHur+v/xs2X3tSMOG5WQis9EcZUNrGC6tQ2eiTMvEc7PUcVS0vYurI59xK0cCNdpY2CxsNUPhcm8R4TT9DNJ98EfLKbFzpQPG5SEyInB4p3E6T/yHY4754eVYWxfZV8qDEWx9eNlEjjm6PuKEguFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIEBJQYc2v6T3JJPafzYT/QPcVG9uxHNPqrRE4IuciXD/vneVmFZcVJDNDDhuBgkvvd2/EBT72K60DI8RQXarav3nS0srLi/oYuPKCEpLEZoSZN7SErsWEGQ2wu7QcWMqnHld//fzz+rbP7aK/drY8m2XOG1/2bl29+uVn/F72+aBuYkmht2h44+/3eo+uHLZMW28nUt3+WPzkzo1yQtq819iE5cWEGTem+ZGl2nz+bTV89xEnxIVdsjNdUvqD8atqQFNcmuY6IxpLTs7MQS3S3bj2Xzh2j1Hz63bFRQ8xacVLyFIl16zpSF90yl4D430e7tEV5BLldN/V7uor4mdiUGQXX392Qa3S7Tx3JqHt49mXeUiWTPp9/FbUNBp63aiKSNtLxZ2uSPIzdCvbV1Rg4nNBLu6C+4EdFbzabFvH826dsXQizV+YlMnF6xibqN2A0E2K1e/jgvyE/qCwkTVJPXatHAbvA65q2G1qi8orGhLUNPBJqQVL9lI22murBQrQY0gH7xpa6dLkG3FXRXzYlLSihfq5htB3Wa40wb557b7XR+0Qa2giTYoFNVO0VIhnu/EJ9qgg8sKTRgzLDFQ9O1xSAVuHvDb9GJuuLb+1NVk18ztdVhhvy97Mb96k2ds+zvbJrt4dsJkL2Z607S04iUPNWqfvHvwI51TMw5yi6xWv023fDEO6gjqTVSqs3oIbsdBplqFeNfHQbp2paUVP9nB6uWh3NJQEICCAE8m6PZQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIMD/3fquyQ9Ims4AAAAASUVORK5CYII=" /><!-- --></p>
</div>
</div>
<div id="references" class="section level2">
<h2>4. References <a name="ref"></a></h2>
<p>Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ. Second-generation PLINK: rising to the challenge of larger and richer datasets. Gigascience. 2015;4:7. <a href="https://gigascience.biomedcentral.com/articles/10.1186/s13742-015-0047-8">link</a></p>
<p>Lee SH, Goddard ME, Wray NR, Visscher PM. A better coefficient of determination for genetic profile analysis. Genet Epidemiol. 2012;36(3):214-24. <a href="http://onlinelibrary.wiley.com/doi/10.1002/gepi.21614/abstract">link</a></p>
<p>Phipson B, Smyth GK. Permutation P-values should never be zero: calculating exact P-values when permutations are randomly drawn. Stat Appl Genet Mol Biol. 2010;9:Article39. <a href="https://www.degruyter.com/view/j/sagmb.2010.9.1/sagmb.2010.9.1.1585/sagmb.2010.9.1.1585.xml">link</a></p>
<p>Pers TH, Timshel P, Hirschhorn JN. SNPsnap: a Web-based tool for identification and annotation of matched SNPs. Bioinformatics. 2015;31(3):418-20. <a href="https://academic.oup.com/bioinformatics/article/31/3/418/2365926">link</a></p>
<p>Wray NR, Lee SH, Mehta D, Vinkhuyzen AA, Dudbridge F, Middeldorp CM. Research review: Polygenic methods and their application to psychiatric traits. J Child Psychol Psychiatry. 2014;55(10):1068-87. <a href="http://onlinelibrary.wiley.com/doi/10.1111/jcpp.12295/abstract">link</a></p>
</div>
<div id="contact" class="section level2">
<h2>5. Contact <a name="email"></a></h2>
<p>javier.costas.costas at sergas.es</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
