#' Generating permutation sets of random SNPs with similar properties than the actual SNP set
#'
#' \code{Generate_permutationSets} generates random sets of SNPs with the same size of the actual SNP set,
#'   aproximately matching the actual SNP set in several properties, such as gene density and distance to nearest
#'   gene, minor allele frequency, and linkage disequilibrium, based on the 1000 genomes European population, taking
#'   data from SNPsnap (Pers et al., 2015).
#'
#'   The extended Major Histocompatibility Complex (chr6, from 25912984 to 33290793 in hg19) was excluded due to its
#'   complex linkage disequilibrium structure.
#'
#'   The SNPs are matched based on deciles except for gene counts, due to the excess of SNPs with low gene counts. Thus,
#'   80.3 percent of SNPs present 0-4 gene counts. The classes to match were 0, 1, 2, 3, 4, and two additional classes, evenly distributed.
#'   Those SNPs not present in the SNPsnap data set or without any matched SNP in the SNPsnap data set are included in
#'   each permutation, to get the same number of SNPs than in the original SNP set.
#'
#' @param input.perm An array of SNPs included in the PRS model, usually, the second element of the list generated
#'   by \code{Calculate_DR2Nagelkerke}.
#' @param sumGWAS A dataframe with a minimum of four columns with names "snpid", "a1", "p","effect", taken from the summary
#'   statistics of the GWAS discovery sample. SNPs in the "snpid" column has to be identified as chr:pos (such as 4:103334).
#'   The "effect" column indicates the effect attributed to allele "a1", as logOR (beta coefficient).
#' @param path.to.plink.files String with the full path to the three PLINK binary files (.bed, .bim, .fam).
#' @param bfile A string indicating the name of the PLINK binary file with data of the GWAS target sample, without extension.
#'   The tmp files generated by the Calculate_DR2Nagelkerke are used by default. If other files are used, the "snpid" column 
#'   has to be identified as chr:pos (such as 4:103334).
#' @param n.perm An integer indicating how many random SNP sets have to be generated.
#'
#' @return A dataframe with the number of random SNP sets especified by \code{n.perm} matching the original
#'  SNP set in several properties.
#'

Generate_permutationSets <- function(input.perm, sumGWAS, path.to.plink.files, bfile = "tmp", n.perm) {

  # Basic check of input files
  if (!"snpid"  %in% colnames(sumGWAS)) stop("No \"snpid\" column in sumGWAS")

  # First, create a subset of the SNPsnap data set with SNPs shared with the summary GWAS and the target GWAS
  bfile.bim <- paste0(path.to.plink.files,bfile,".bim")
  b.file <- read.table(bfile.bim, header = F)
  # Replace SNP_ID by chr:position
  b.file$V2 <- paste(b.file$V1,b.file$V4,sep=":")
  SNPsnap.subsets.co <- subset(SNPsnap.subsets, SNPsnap.subsets$snpid %in% sumGWAS$snpid)
  SNPsnap.subsets.common <- subset(SNPsnap.subsets.co, SNPsnap.subsets.co$snpid %in% b.file$V2)

  # Identify input SNPs not in SNPsnap
  absent <- subset(input.perm, !input.perm %in% SNPsnap.subsets.common$snpid)
  if (length(absent) == 1) {
    message("\nThere are no data on SNP properties for one SNP from the SNP set")
  }
  if (length(absent) > 1) {
    message("\nThere are no data on SNP properties for ", length(absent), " SNPs from the SNP set")
  }
  present <- subset(input.perm, input.perm %in% SNPsnap.subsets.common$snpid)

  #Get information of input SNPs
  datos.input <- subset(SNPsnap.subsets.common, SNPsnap.subsets.common$snpid %in% present)

  #Create an empty dataframe to store the permutations
  permut <- data.frame(matrix(NA, nrow = length(input.perm), ncol = n.perm))

  #Add absent SNPs
  if (length(absent) > 0) {
    for (i in 1:length(absent)) {
    permut[i, ] <- rep(absent[i], n.perm)
    }
  }
  #Add matched SNPs
  for (i in 1:dim(datos.input)[1]) {
    pool <- subset(SNPsnap.subsets.common, SNPsnap.subsets.common$subset == datos.input$subset[i])
    #Fill data.frame
    selected <- sample(x = pool$snpid, size = n.perm, replace = T)
    permut[length(absent) + i, ] <- selected
  }

 return(permut)

}
